### ISSUE TITLE
Log Structured Merge Trees - ben stopford

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6847]
Log Structured Merge Trees - ben stopford<br>
<br>
https://ift.tt/17BnJht<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<div>
<p>It&rsquo;s nearly a decade since Google released its &lsquo;Big Table&rsquo; paper. One of the many cool aspects of that paper&nbsp;was the file organisation it uses. The approach is more generally known as the Log Structured Merge Tree, after <a href="http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.44.2782&amp;rep=rep1&amp;type=pdf">this</a> 1996 paper, although the algorithm described there differs quite significantly from most real-world implementations.</p>
<p>LSM is now used in a number of&nbsp;products as the main file organisation strategy. HBase, Cassandra, LevelDB, SQLite, even MongoDB 3.0 comes with an optional LSM engine, after it&rsquo;s acquisition of Wired Tiger.</p>
<p>What makes LSM trees interesting is their&nbsp;departure from binary tree style&nbsp;file organisations that have dominated the space for decades. LSM seems almost counter intuitive&nbsp;when you first look at&nbsp;it, only making sense when you closely consider how files work in modern, memory heavy systems.</p>
<h3>Some Background</h3>
<p>In a nutshell LSM trees are designed to provide better write throughput than traditional B+ tree or <a href="http://en.wikipedia.org/wiki/ISAM">ISAM</a> approaches. They do this by removing the need to perform dispersed,&nbsp;update-in-place operations.</p>
<p><span><img src="http://www.benstopford.com/wp-content/uploads/2015/02/ChartGo-300x267.png" alt="ChartGo" width="300" class="wp-image-3554" height="267">So why is this a good idea? At its core it&rsquo;s the old problem of&nbsp;disks being slow for random operations,&nbsp;but fast when accessed sequentially. A gulf exists between these two types of access, regardless of whether the disk is magnetic or solid state or even, although to a lesser extent, main memory.</span></p>
<p><span>The figures in this ACM report <a href="http://queue.acm.org/detail.cfm?id=1563874">here</a>/<a href="http://deliveryimages.acm.org/10.1145/1570000/1563874/jacobs3.jpg">here</a>&nbsp;make the point well. They show that,&nbsp;somewhat counter intuitively, sequential disk access is faster than randomly accessing&nbsp;main memory. More relevantly they also show sequential access to disk, be it magnetic or SSD, to be at least three orders of magnitude faster than&nbsp;random IO. This means random operations are to be avoided. Sequential access is&nbsp;well worth designing for.</span></p>
<p><span>So with this in mind lets consider a little thought experiment: if we are interested in write throughput, what is the best method to use?&nbsp;A good&nbsp;starting point is&nbsp;to&nbsp;simply append data to a file. This approach, often termed&nbsp;logging, journaling or a heap file, is fully sequential so provides very fast write performance equivalent to theoretical disk speeds (typically 200-300MB/s per drive).</span></p>
<p>Benefiting from&nbsp;both simplicity and performance log/journal&nbsp;based approaches have rightfully become <a href="http://engineering.linkedin.com/distributed-systems/log-what-every-software-engineer-should-know-about-real-time-datas-unifying">popular</a> in many big data tools. Yet they have an obvious downside. Reading arbitrary data from a log&nbsp;will be far more time consuming than writing to it, involving a reverse chronological scan, until the required key is found.</p>
<p>This means&nbsp;logs are only really applicable to&nbsp;simple workloads, where&nbsp;data is either accessed in its entirety, as in the write-ahead log of most databases, or by a known offset, as in simple messaging products like Kafka.</p>
<p>So we need more than just a journal to efficiently perform more&nbsp;complex read workloads like key based access or a range search. Broadly speaking there are four&nbsp;approaches that can help us here: binary search, hash, B+ or external.</p>
<ol>
<li>Search Sorted File: save data to a file,&nbsp;sorted by key. If data has defined widths use Binary search. If not&nbsp;use a&nbsp;page index + scan.</li>
<li>Hash: split the data into buckets using a hash function, which can later be used to direct&nbsp;reads.</li>
<li><span>B+: use navigable file organisation such as a B+ tree,&nbsp;<a href="http://en.wikipedia.org/wiki/ISAM%20">ISAM</a>&nbsp;etc.&nbsp;</span></li>
<li>External file: leave the data as a log/heap and create a separate&nbsp;hash or tree index into it.</li>
</ol>
<p>All these approaches improve read performance significantly ( n-&gt;O(log(n)) in most). Alas these structures add order and that order impedes write performance, so our high speed&nbsp;journal file is lost along the way. You can&rsquo;t have your cake and eat it I guess.</p>
<p><img src="http://www.benstopford.com/wp-content/uploads/2015/02/tree-2-1024x293.jpg" alt="tree (2)" width="716" class="wp-image-3611" height="205"></p>
<p>An insight that is worth making is that all&nbsp;four of the above options impose some form of&nbsp;overarching structure on the data.</p>
<p>Data is deliberately and specifically placed around the file system so the index can quickly find it again&nbsp;later. It&rsquo;s this structure that makes navigation quick. Alas the&nbsp;structure must of course be honoured as data is written. This is where we start to degrade write performance by adding in random disk access.</p>
<p>There are a couple of specific issues. Two IOs are needed for each write, one to read the page and one to write it back. This wasn&rsquo;t the case with our log/journal file which could do it in one.</p>
<p>Worse though, we now need to update the structure of the hash or B+ index. This means updating specific parts of the file system. This is known as update-in-place and requires slow, random IO.&nbsp;<i>This point is important:</i>&nbsp;in-place approaches like this scatter-gun the file system performing update-in-place*. This is limiting.</p>
<p>One common solution is to use approach (4) A index into a journal &ndash; but keep the index in memory. So, for example, a Hash Table can be&nbsp;used to map keys to the position (offset) of the latest value in a journal file (log). This approach actually works pretty&nbsp;well as&nbsp;it compartmentalises random IO to something relatively small: the key-to-offset mapping, held in memory. Looking up a value is then&nbsp;only a single IO.</p>
<p>On the other hand there are scalability limits, particularly if you have lots of small values. If your values were just say simple numbers then the index would be larger than the data file itself. Despite this the&nbsp;pattern is a sensible compromise which is used in many products from Riak through to Oracle Coherence.</p>
<p>So this brings us on to Log Structured Merge Trees. LSMs take a different approach to the four above. They can be fully disk-centric, requiring little in memory storage for efficiency, but also hang&nbsp;onto much of the write performance we would tie to a simple journal file. The one downside is slightly poorer read performance when compared to say a B+Tree.</p>
<p>In essence they do everything they can to make disk access&nbsp;sequential. No scatter-guns here!</p>
<p>*A number of tree structures exist which&nbsp;do not require update-in-place. Most popular is the&nbsp;<a href="http://www.bzero.se/ldapd/btree.html">append-only Btree</a>, also know as the copy-on-write tree. These work by overwriting&nbsp;the tree structure, sequentially,&nbsp;at&nbsp;the end of&nbsp;the file each&nbsp;time a write occurs. Relevant parts of the old tree structure, including the top level node, are orphaned. Through this method&nbsp;update-in-place is avoided as the tree sequentially redefines itself over time. This method does however come&nbsp;at the cost: rewriting the structure on every write is&nbsp;verbose. It&nbsp;creates a significant amount of write amplification which is a&nbsp;downside unto itself.</p>
<p><a href="http://www.benstopford.com/wp-content/uploads/2015/02/Journal6.png"><img src="http://www.benstopford.com/wp-content/uploads/2015/02/Journal6-1024x503.png" class="wp-image-3638"></a></p>
<h3>The Base LSM Algorithm</h3>
<p>Conceptually the base LSM tree is fairly&nbsp;simple. Instead of having one big index structure (which will either&nbsp;scatter-gun the file system or add significant write amplification) batches of writes are saved, sequentially, to a set of smaller index files. So each&nbsp;file contains a batch of changes covering a short period of time. Each file&nbsp;is sorted before it is written so searching it later will be fast. Files are immutable; they are&nbsp;never updated. New updates go into new files. Reads inspect all files.&nbsp;Periodically files&nbsp;are merged together to&nbsp;keep the number of files down.</p>
<p>Lets look at this in a little more detail. When updates arrive&nbsp;they are added to an in-memory buffer, which is usually held as a tree (Red-Black etc) to preserve&nbsp;key-ordering. This &lsquo;memtable&rsquo; is replicated on disk as a write-ahead-log in most implementations, simply for recovery purposes. When the memtable fills the sorted data&nbsp;is flushed to a new&nbsp;file on disk. This process repeats as more and more writes come in. Importantly the system is only doing sequential IO as files are not edited. New entries&nbsp;or edits simply create successive files (see fig above).</p>
<p>So as more data comes into the system, more and more of these immutable, ordered files are created. Each one&nbsp;representing a small, chronological subset of changes, held sorted.</p>
<p><span>As old files are not updated duplicate entries are created to supersede previous records&nbsp;(or removal markers). This creates some redundancy initially.</span></p>
<p><span>Periodically the system performs a <em>compaction.&nbsp;</em>Compaction selects multiple files and merges them together, removing any duplicated updates or deletions (more on how this works later). This is important both to remove the aforementioned redundancy but, more importantly, to keep a handle on the read performance which degrades as the number of files increases. Thankfully, because the files are sorted, the process of merging the files is quite efficient.&nbsp;</span></p>
<p>When a read operation is requested the system first checks the in memory buffer (memtable). If the key is not found&nbsp;the various files will be inspected one by one, in reverse chronological order, until the key is found. Each file is held sorted so it is navigable. However reads will become slower and slower as the number of files increases, as each one needs to be inspected. This is a problem.</p>
<p><span>So&nbsp;reads in LSM trees are slower than their in-place brethren. Fortunately there are&nbsp;a couple of tricks which can make the pattern performant. The most common approach is to&nbsp;hold a&nbsp;page-index in memory. This provides a lookup which gets you &lsquo;close&rsquo; to your target key. You scan from there as the data is sorted. <a href="http://leveldb.googlecode.com/svn/trunk/doc/impl.html">LevelDB</a>, <a href="http://rocksdb.org/" target="_blank">RocksDB</a>&nbsp;and BigTable do this with a block-index held at the end of each file. This often works better than straight binary search as it allows the use of variable length fields and is&nbsp;better suited to&nbsp;compressed data.&nbsp;</span></p>
<p><span>Even with per-file indexes read operations will still slow as the number of files increases. This is kept in check by periodically merging files together. Such compactions keep the number of files,&nbsp;and hence&nbsp;read performance, within acceptable bounds.<br></span></p>
<p>Even with compaction reads will still need to visit many files. Most implementations void this through the use of a&nbsp;<a href="http://en.wikipedia.org/wiki/Bloom_filter">Bloom filter</a>. Bloom filters are a memory efficient way of working out whether a file contains a key.</p>
<p>So from a &lsquo;write&rsquo; perspective; all writes are batched up and written <em><strong>only</strong></em> in sequential chunks. There is an additional, periodic IO penalty from compaction rounds. Reads however have&nbsp;the potential to touch a&nbsp;large number of files when looking up a single row (i.e. scatter-gun on read). This is simply the way the algorithm works. We&rsquo;re trading random IO on write for random IO on read. This trade off is&nbsp;sensible if we can use software tricks like bloom filters or hardware tricks like large file caches to optimise read performance.</p>
<p><a href="http://www.benstopford.com/wp-content/uploads/2015/02/Journal2.51.png"><img alt="Journal3" src="http://www.benstopford.com/wp-content/uploads/2015/02/Journal31-1024x629.png" class="wp-image-3643"><br></a></p>
<h3>Basic Compaction</h3>
<p>To keep LSM reads relatively fast it&rsquo;s important to&nbsp;manage-down the number of files, so lets look more deeply at compaction. The process is a bit like generational garbage collection:</p>
<p>When a certain number of files have been created, say five&nbsp;files, each with 10 rows, they are merged into a&nbsp;single file, with 50 rows (or maybe slightly less) .</p>
<p>This process continues with more 10 row files being created. These are merged into 50 row files every time the fifth file fills up.</p>
<p>Eventually there are five 50 row files. At this point the five&nbsp;50 row files are merged into one 250&nbsp;row file. The process continues creating larger and larger files. See fig.</p>
<p><span>The aforementioned issue with this general approach is the large number of files that are created: all must be searched, individually, to read a result (at least in the worst case).</span></p>
<p><img alt="Journal2.5" src="http://www.benstopford.com/wp-content/uploads/2015/02/Journal2.51-1024x514.png" class="wp-image-3642"></p>
<h3>Levelled Compaction</h3>
<p><span>Newer implementations, such as those in&nbsp;LevelDB, RocksDB and Cassandra, address this problem by implementing a level-based, rather than size-based, approach to compaction. This reduces the number of files that must be consulted for&nbsp;the worst case read, as well as reducing the relative impact of a single compaction.&nbsp;</span></p>
<p><span>This level-based approach has two key differences compared to the base approach above:</span></p>
<p>1. Each level can contain a number of files and is guaranteed, as a whole, to not have overlapping keys within it. That is to say the keys are partitioned across the available files.&nbsp;Thus&nbsp;to find a key in a certain level only one file needs to be consulted.</p>
<p>The first level is a special case where the above property does not hold. Keys can span&nbsp;multiple files.</p>
<p>2. Files are merged into upper levels&nbsp;one file at a time. As a level fills, a single file is plucked from it and merged into the level above creating space for more data to be added. This is slightly different to the base-approach where&nbsp;several similarly sized files are merged into a single,&nbsp;larger one.</p>
<p>These changes mean&nbsp;the level-based&nbsp;approach&nbsp;spreads the impact of compaction over time as well as requiring less total space. It also has better read performance. However the total IO is higher for most workloads meaning some of the simpler write-oriented workloads will not see benefit.</p>
<h3>Summarising</h3>
<p>So LSM trees sit in the middle-ground between a journal/log file and a traditional single-fixed-index such as a B+ tree or Hash index. They provide&nbsp;a mechanism for managing a set of smaller, individual index files.</p>
<p>By managing a group of indexes, rather than a single one, the LSM method&nbsp;trades&nbsp;the&nbsp;expensive random IO associated with update-in-place in B+ or Hash indexes for fast, sequential IO.</p>
<p>The price being paid is&nbsp;that reads have to address a large number of index files rather than just the one. Also there is additional IO cost for compaction.</p>
<p>If that&rsquo;s still a little murky there are some other good&nbsp;descriptions&nbsp;<a href="http://leveldb.googlecode.com/svn/trunk/doc/impl.html">here</a> and <a href="http://www.datastax.com/dev/blog/leveled-compaction-in-apache-cassandra">here</a>.</p>
<h3>Thoughts on the LSM approach</h3>
<p><span>So are LSM approaches really better than traditional single-tree&nbsp;based ones?</span></p>
<p><span>We&rsquo;ve seen that LSM&rsquo;s have better write performance albeit a cost. LSM has some other benefits though. The SSTables (the sorted files) a&nbsp;</span>LSM tree creates are immutable. This makes the locking semantics over them much simpler. Generally the only resource that is contended is the memtable. This is in contrast to singular trees which require&nbsp;elaborate locking mechanisms to manage changes at different levels.</p>
<p><span>So ultimately the question is likely to be about how write-oriented expected&nbsp;workloads are. If you care about write performance the savings LSM gives are likely to be&nbsp;a big deal. The big internet companies seem pretty settled on this subject. Yahoo, for <a href="http://www.eecs.harvard.edu/~margo/cs165/papers/gp-lsm.pdf">example</a>, reports a steady progression from read-heavy to read-write workloads, driven largely by the increased ingestion of event logs and mobile data. Many traditional database products still seem to favour more read-optimised file structures though.&nbsp;</span></p>
<p><span>As with Log Structured file systems [see footnote] the key argument stems from the increasing availability of memory. With more memory available reads are naturally optimised through&nbsp;large file caches provided by the operating system. Write performance (which memory doesn&rsquo;t improve with more) thus becomes&nbsp;the dominant concern. So put another way, hardware advances are doing more for read performance than they are&nbsp;for writes. Thus it makes sense to select a write-optimised file structure.</span></p>
<p><span>Certainly LSM implementations such as LevelDB and Cassandra regularly provide better write performance than single-tree&nbsp;based approaches (<a href="http://symas.com/mdb/microbench/">here</a> and <a href="https://s3-eu-west-1.amazonaws.com/benstopford/nosql-comp.pdf">here</a> respectively).</span></p>
<h3>Beyond Levelled&nbsp;LSM</h3>
<p><span>There has been a fair bit of further work building on the LSM approach. Yahoo developed a system called <a href="http://highscalability.com/blog/2009/8/8/yahoos-pnuts-database-too-hot-too-cold-or-just-right.html">Pnuts</a>&nbsp;which combines LSM with B trees&nbsp;and <a href="http://www.eecs.harvard.edu/~margo/cs165/papers/gp-lsm.pdf">demonstrates</a>&nbsp;better performance.&nbsp;I haven&rsquo;t seen&nbsp;openly available implementations of this algorithm though. IBM and Google have done more&nbsp;recent work in&nbsp;a similar vein, albeit via a different <a href="http://researcher.ibm.com/researcher/files/us-wtan/DiffIndex-EDBT14-CR.pdf">path</a>. There are also related&nbsp;approaches which have similar properties but retain an overarching structure. These include <a href="http://highscalability.com/blog/2014/8/6/tokutek-white-paper-a-comparison-of-log-structured-merge-lsm.html">Fractal Trees</a>&nbsp;and <a href="https://www.cs.ox.ac.uk/people/andy.twigg/pubs/2011-hotstorage.pdf">Stratified Trees</a>. &nbsp;</span></p>
<p>This is of course just one alternative. Databases utilise a huge range of subtly different options. An increasing number of databases offer pluggable&nbsp;engines for different workloads.&nbsp;<a href="http://parquet.incubator.apache.org/">Parquet</a> is a popular alternative for HDFS and pushes in pretty much the opposite direction (aggregation performance via&nbsp;a columnar format). MySQL &nbsp;has a storage abstraction which is pluggable with&nbsp;a number of different engines such as&nbsp;<a href="http://www.tokutek.com/">Toku</a>&lsquo;s fractal tree based&nbsp;index. This is also available for&nbsp;MongoDB. Mongo 3.0 includes the <a href="https://github.com/wiredtiger/wiredtiger/wiki/LevelDB-Benchmark">Wired Tiger</a> engine which provides both&nbsp;B+ &amp; LSM approaches along with the legacy engine. Many relational databases have configurable&nbsp;index structures that utilise different file organisations.</p>
<p>It&rsquo;s also worth&nbsp;considering the hardware being used. Expensive solid state disks, like FusionIO, have better random write performance. This suits update-in-place approaches. Cheaper SSDs and mechanical drives are better suited to LSM. LSM&rsquo;s avoid the small random access patters that thrash SSDs into oblivion**.</p>
<p>LSM is not without it critics though. It&rsquo;s biggest problem, like GC, is the collection phases and the effect they have on precious IO. There is an interesting discussion of some of these on <a href="https://news.ycombinator.com/item?id=9145197">this</a> hacker news thread.</p>
<p>So if you&rsquo;re looking at data products, be it BDB&nbsp;vs. LevelDb, Cassandra vs.&nbsp;MongoDb you may tie some proportion of their relative performance back to the file structures they use. <a href="https://github.com/benstopford/awesome-db-benchmarks">Measurements</a> appear to back this philosophy. Certainly it&rsquo;s worth being aware&nbsp;of the performance tradeoffs being selected&nbsp;by the systems you use.</p>
<hr>
<hr>
<p>**In SSDs each write incurs a clear-rewrite cycle for a&nbsp;whole 512K block. Thus small writes can induce a disproportionate amount of churn on the drive. With fixed limits on block rewrites this can significantly affect their life.</p>
<h3><span>Further Reading</span></h3>
<ul>
<li>There is a nice introductory post <a href="https://www.igvita.com/2012/02/06/sstable-and-log-structured-storage-leveldb/">here</a>.</li>
<li>The LSM description in this&nbsp;<a href="http://www.eecs.harvard.edu/~margo/cs165/papers/gp-lsm.pdf">paper</a>&nbsp;is great and it also discusses interesting extensions.</li>
<li>These three posts provide a holistic coverage of the algorithm:&nbsp;<a href="http://leveldb.googlecode.com/svn/trunk/doc/impl.html">here</a>,&nbsp;<a href="http://www.datastax.com/dev/blog/leveled-compaction-in-apache-cassandra">here</a>&nbsp;and <a href="http://www.quora.com/How-does-the-Log-Structured-Merge-Tree-work">here</a>.</li>
<li>The original Log Structured Merge Tree paper <a href="http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.44.2782&amp;rep=rep1&amp;type=pdf">here</a>. It&nbsp;is a little&nbsp;hard to follow in my opinion.</li>
<li>The&nbsp;Big Table paper&nbsp;<a href="http://static.googleusercontent.com/media/research.google.com/en//archive/bigtable-osdi06.pdf">here</a>&nbsp;is excellent.</li>
<li>
<a href="http://highscalability.com/blog/2014/8/6/tokutek-white-paper-a-comparison-of-log-structured-merge-lsm.html">LSM vs Fractal Trees</a> on High Scalability.</li>
<li>Recent work on <a href="http://researcher.ibm.com/researcher/files/us-wtan/DiffIndex-EDBT14-CR.pdf">Diff-Index</a> which builds on the LSM concept.</li>
<li>
<a href="http://blog.empathybox.com/post/24415262152/ssds-and-distributed-data-systems">Jay</a>&nbsp;on SSDs and the benefits of LSM</li>
<li>Interesting <a href="https://news.ycombinator.com/item?id=9145197">discussion</a> on hackernews regarding index structures.</li>
</ul>
<h3><span>Footnote on log structured file systems</span></h3>
<p><span>Other than the name, and a focus on write throughput, there isn&rsquo;t that much relation between LSM and log structured file systems as far as I can see.</span></p>
<p><span>Regular filesystems used today tend to be &lsquo;Journaling&rsquo;, for example ext3, ext4, HFS etc are&nbsp;tree-based approaches. A fixed height tree of inodes represent the directory structure and a journal is used to protect against failure conditions. In these implementations the journal is logical, meaning it only internal&nbsp;metadata will be journaled. This is for performance reasons.</span></p>
<p><span>Log structured file systems are widely used on flash media as they have less write amplification. They are getting more press too as file caching starts to dominate read workloads in more general situations and write performance is becoming more critical.</span></p>
<p><span>In log structured file systems data is written only once, directly to a journal which is&nbsp;</span><span>represented</span><span>&nbsp;as a chronologically advancing buffer. The buffer is garbage collected periodically to remove redundant writes. Like LSM&rsquo;s the log structured file system will write faster, but read slower than its dual-writing, tree based counterpart. Again this is&nbsp;acceptable where there is lots of RAM available to feed the file cache&nbsp;or the media doesn&rsquo;t deal well with update in place, as is the case with&nbsp;flash.</span></p>
<p>Posted on February 14th, 2015 in <a rel="category tag" href="http://www.benstopford.com/category/analysis/">Analysis and Opinion</a>, <a rel="category tag" href="http://www.benstopford.com/category/bigdata/">Big Data</a>, <a rel="category tag" href="http://www.benstopford.com/category/blogroll/">Blog</a>, <a rel="category tag" href="http://www.benstopford.com/category/top4/">Top4</a></p>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via ben stopford https://ift.tt/JBeye6<br>
<br>
June 18, 2020 at 08:54PM
